### ISSUE TITLE
QML modules shipped in AppImage are not recognized

### ISSUE THREAD

**Issue by petrmanek** [probonopd/linuxdeployqt#586]
I am distributing a Qt Quick app as AppImage using linuxdeployqt. Recently, I upgraded my distribution stack to Ubuntu Focal from Ubuntu Bionic (following its support sunset). From that point onward, I am unable to run any AppImages produced.

My app always fails launch with the following error message

```
QQmlComponent: Component is not ready
qrc:/gui/qml/SessionWindow.qml:1:1: Failed to extract plugin meta data from '/tmp/.mount_Vm6MJb/usr/qml/QtQml/WorkerScript/libworkerscriptplugin.so': '/tmp/.mount_Vm6MJb/usr/qml/QtQml/WorkerScript/libworkerscriptplugin.so' is not a Qt plugin (metadata not found)
```

If I extract the AppImage and go to `usr/qml/QtQml/WorkerScript` and run `file libworkerscriptplugin.so`, I see the following output, which looks sane to me:

```
libworkerscriptplugin.so: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=56b7002ffb32efabb1b116bb6ba4fab51d7946d5, not stripped
```

Furthermore, the `ldd libworkerscriptplugin.so` command yields the following output:

```
	linux-vdso.so.1 (0x00007ffd38db3000)
	libQt6QmlWorkerScript.so.6 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libQt6QmlWorkerScript.so.6 (0x00007fc5b4698000)
	libQt6Qml.so.6 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libQt6Qml.so.6 (0x00007fc5b4000000)
	libQt6Network.so.6 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libQt6Network.so.6 (0x00007fc5b3e60000)
	libQt6Core.so.6 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libQt6Core.so.6 (0x00007fc5b3800000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fc5b467c000)
	libstdc++.so.6 => /lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007fc5b3400000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fc5b3719000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007fc5b465c000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fc5b3000000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fc5b46b4000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fc5b4657000)
	libgssapi_krb5.so.2 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libgssapi_krb5.so.2 (0x00007fc5b4605000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fc5b45e9000)
	libicui18n.so.66 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libicui18n.so.66 (0x00007fc5b2c00000)
	libicuuc.so.66 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libicuuc.so.66 (0x00007fc5b2800000)
	libicudata.so.66 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libicudata.so.66 (0x00007fc5b0c00000)
	libglib-2.0.so.0 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libglib-2.0.so.0 (0x00007fc5b32cc000)
	libdouble-conversion.so.3 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libdouble-conversion.so.3 (0x00007fc5b45cf000)
	libpcre2-16.so.0 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libpcre2-16.so.0 (0x00007fc5b3ddb000)
	libgthread-2.0.so.0 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libgthread-2.0.so.0 (0x00007fc5b45c9000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fc5b45c4000)
	libkrb5.so.3 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libkrb5.so.3 (0x00007fc5b3637000)
	libk5crypto.so.3 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libk5crypto.so.3 (0x00007fc5b3da9000)
	libcom_err.so.2 => /lib/x86_64-linux-gnu/libcom_err.so.2 (0x00007fc5b45bc000)
	libkrb5support.so.0 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libkrb5support.so.0 (0x00007fc5b45ac000)
	libpcre.so.3 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libpcre.so.3 (0x00007fc5b3258000)
	libkeyutils.so.1 => /path/redacted/squashfs-root/usr/qml/QtQml/WorkerScript/./../../../lib/libkeyutils.so.1 (0x00007fc5b45a2000)
	libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007fc5b3d95000)
```

In the listing above, you can see that all dependencies are accounted for, and mostly provided from within AppImage's file tree, as they should be.

Finally, running `readelf -a libworkerscript.so` yields voluminous output, at the end of which there is the following entry:

```
[...]

Displaying notes found in: .note.qt.metadata
  Owner                Data size 	Description
  qt-project!          0x00000068	Unknown note type: (0x74510001)
   description data: 01 06 04 01 bf 02 78 2e 6f 72 67 2e 71 74 2d 70 72 6f 6a 65 63 74 2e 51 74 2e 51 51 6d 6c 45 6e 67 69 6e 65 45 78 74 65 6e 73 69 6f 6e 49 6e 74 65 72 66 61 63 65 03 77 51 74 51 6d 6c 57 6f 72 6b 65 72 53 63 72 69 70 74 50 6c 75 67 69 6e 63 75 72 69 81 72 51 74 51 6d 6c 2e 57 6f 72 6b 65 72 53 63 72 69 70 74 ff 
```

To me, this convincingly refutes the Qt error message that causes my program to crash -- the `libworkerscript.so` file clearly contains Qt metadata in that very ELF note. Could there be something else causing this message to appear? Perhaps a file that failed to be included by `linuxdeployqt`? Have you seen a similar issue before?

For due diligence I should probably mention that I performed another two changes which may have had impact:

* Upgraded cmake from 3.16 to 3.27
* Added `-feature-optimize_full` to my Qt Everywhere build

I am using Qt Everywhere v6.4.3 built on the latest Ubuntu Focal with linuxdeployqt `e83eeb7b77a8982a94ddb9de493049e15bc687cb` built from source. When packaging my software, I invoke `linuxdeployqt` with the `-appimage` option and `-extra-plugins` that are equal to all plugins in `/usr/plugins`, as well as `-qmldir` options for every QML directory in my project.

Thanks for your thoughts! P.

**Comment by probonopd** [comment]
Thanks for reporting. Unfortunately I don't have a ready solution. But what happens if you use the Qt from https://launchpad.net/~beineri/+archive/ubuntu/opt-qt-5.15.4-focal?

Also note that linuxdeployqt does not really support Qt6. It never really has been fully ported or tested there so far.

**Comment by petrmanek** [comment]
> But what happens if you use the Qt from https://launchpad.net/~beineri/+archive/ubuntu/opt-qt-5.15.4-focal?

I will try that and get back to you.

> Also note that linuxdeployqt does not really support Qt6 (even though it may or may not work). It never really has been fully ported or tested there so far.

Acknowledged. For the record, I have been using linuxdeployqt with Qt 6 apps for over a year at this point, and it worked rather nicely until now. This is just an edge case that will have to be considered if/when Qt 6 support is fully implemented.

**Comment by petrmanek** [comment]
One more note, just tested with a Qt Everywhere without `-feature-optimize_full`, and am getting the same result.

**Comment by petrmanek** [comment]
I think that I may have solved the issue. It appears that it was not caused by linuxdeployqt somehow mangling the packaged binaries, but rather by a Qt distribution that was built without a correctly configured locale. I did not really pay too much attention to the following messages that popped up throughout the build of Qt itself:

```
Detected system locale encoding (ANSI_X3.4-1968, locale "C") is not UTF-8.
Qt shall use a UTF-8 locale ("C.UTF-8") instead. If this causes problems,
reconfigure your locale. See the locale(1) manual for more information.
```

It appears that this has a direct effect on the metadata that is generated. Even though Qt builds along with this warning, all of its distributed plugins fail to load with the error that I reported in my first message. Perhaps there is some ambiguity about the format, in which the magic string searched for by `QElfParser` is serialized.

In my case, the solution was to set the following environment variables before Qt was compiled:

```
LC_ALL=en_US.UTF-8
LANG=en_US.UTF-8
LANGUAGE=en_US.UTF-8
```

This made the locale warning disappear. Furthermore, if I attempt to load `libworkerscriptplugin.so`, I am able to extract the following metadata (even when opened from another x86_64 PC):

```
qt.core.plugin.loader: Found metadata in lib /tmp/libworkerscriptplugin.so, metadata=
{
    "IID": "org.qt-project.Qt.QQmlEngineExtensionInterface",
    "archlevel": 1,
    "className": "QtQmlWorkerScriptPlugin",
    "debug": false,
    "uri": [
        "QtQml.WorkerScript"
    ],
    "version": 394240
}
```

Curiously enough, the output of `readelf` has not changed in any significant way.

I am now attempting to verify that this solution works on large scale. I will post my result here.

**Comment by petrmanek** [comment]
I have isolated the versions of `libworkerscriptplugin.so` that seem to work (have Qt metadata) and not work (also have Qt metadata, which does not seem to be found by `QElfParser`). Here they are for future reference: [libworkerscriptplugin.tar.gz](https://github.com/probonopd/linuxdeployqt/files/12763451/libworkerscriptplugin.tar.gz)

After rebuilding my entire Qt chain, I can now safely say that **implicitly configured default locale is not the culprit here**. I have used the attached files to track the journey from `libworkerscriptplugin.so` all the way from Qt's installed location into my AppImage. It seems that at some point `linuxdeployqt` calls `patchelf --set-rpath` on all ELF files -- this appears to cause damage to embedded Qt metadata to the extent that it cannot be read back from the resulting file.

Here is a minimal example that I distilled after hours of testing:

```bash
cp /usr/qml/QtQml/WorkerScript/libworkerscriptplugin.so .
ls -l libworkerscriptplugin.so
MY_RPATH=$(patchelf --print-rpath libworkerscriptplugin.so)
patchelf --set-rpath "$MY_RPATH:\$ORIGIN/../asd" libworkerscriptplugin.so
ls -l libworkerscriptplugin.so
```

The script above copies `libworkerscriptplugin.so`, reads its rpath and extends it with `$ORIGIN/../asd`. This is there just to perform some meaningful change, otherwise `patchelf` is smart enough not to do anything. Unfortunately, after this the Qt metadata can no longer be parsed from the binary.

**Comment by petrmanek** [comment]
The latest version of Ubuntu Focal, where I am running `linuxdeployqt`, uses patchelf v0.10. Having repeated the test with the [oldest online provided binaries (v0.14.3)](https://github.com/NixOS/patchelf/releases/tag/0.14.3), I can report that with patchelf v0.14.3 the bug is no longer present. I would say that this is a good basis for bisection: somewhere between v0.10 and v0.14.3 the bug must have been corrected.

For me, this means that I will simply use a locally built version of patchelf instead of the one from Ubuntu package repository.

**Comment by petrmanek** [comment]
I just verified that using a newer version of `patchelf` solved the issue for me. So just to summarize, the culprit was not `linuxdeployqt`, it was not the locale used when building Qt framework, or any other thing that I changed recently. Instead, it was `patchelf` which messed up the binaries in such a way that later on Qt could not load them properly.

It may be a good idea to try to detect this issue in `linuxdeployqt` and raise a warning when it happens, otherwise people are going to spend long hours pulling their hair over this issue, like I did.

Here is a quick and easy way to detect Qt metadata loss that was useful during my debugging:

```cpp
QPluginLoader plugin("/path/to/tested/file/libworkerscriptplugin.so");
qDebug() << plugin.metaData();

// When metadata is lost, the following is printed
// QJsonObject()
//
// ... otherwise you should see something like this:
// QJsonObject({"IID":"org.qt-project.Qt.QQmlEngineExtensionInterface","archlevel":1,"className":"QtQmlWorkerScriptPlugin","debug":false,"uri":["QtQml.WorkerScript"],"version":394240})
```
